source/dem/disable_particle_contact.cc

```diff
@@ -0,0 +1,335 @@
+          status_to_cell[mobility_status::mobile].insert(*cell);
+
+          // Remove cell from cell set
+          cell = active_ghost_cells.erase(cell);
+
+          // Assign mobile status to nodes
+          for (auto dof_index : local_dofs_indices)
+            {
+              // Don't overwrite empty nodes
+              mobility_at_nodes(dof_index) =
+                std::max((int)mobility_status::mobile,
+                         mobility_at_nodes[dof_index]);
+            }
+        }
+      else
+        {
+          ++cell;
+        }
+    }
+
+  mobility_at_nodes.update_ghost_values();
+
+  // Layer of mobile cells over mobile cells
+  for (auto cell = active_ghost_cells.begin();
+       cell != active_ghost_cells.end();)
+    {
+      std::vector<types::global_dof_index> local_dofs_indices(dofs_per_cell);
+      (*cell)->get_dof_indices(local_dofs_indices);
+      bool has_mobile_nodes = false;
+
+      // Loop over nodes of cell
+      for (auto dof_index : local_dofs_indices)
+        {
+          // Check if node is mobile and assign mobile status to
+          // the cell
+          if (mobility_at_nodes[dof_index] == mobility_status::mobile)
+            {
+              status_to_cell[mobility_status::mobile].insert(*cell);
+
+              // Remove cell from cell set
+              cell             = active_ghost_cells.erase(cell);
+              has_mobile_nodes = true;
+
+              // Assign active status to nodes except mobile because
+              // this will cause to propagate the mobile status to the
+              // neighbors in this loop since the mobility check at node
+              // is executed in the same container that we assign new
+              // mobility status
+              // Since mobile = 2 > active = 1, this step won't overwrite
+              // mobile nodes
+              for (auto dof_index : local_dofs_indices)
+                {
+                  mobility_at_nodes[dof_index] =
+                    std::max((int)mobility_status::active,
+                             mobility_at_nodes[dof_index]);
+                }
+              break;
+            }
+        }
+
+      if (!has_mobile_nodes)
+        {
+          ++cell;
+        }
+    }
+
+  mobility_at_nodes.update_ghost_values();
+
+  // Layer of neighbor of mobile cells
+  for (auto cell = active_ghost_cells.begin();
+       cell != active_ghost_cells.end();)
+    {
+      std::vector<types::global_dof_index> local_dofs_indices(dofs_per_cell);
+      (*cell)->get_dof_indices(local_dofs_indices);
+
+      bool has_active_nodes = false;
+      bool has_mobile_nodes = false;
+
+      // Loop over nodes of cell and check if cell has active or mobile
+      // nodes
+      // Active nodes and no mobiles nodes means that the cell is a neighbor
+      // of the layer of cell over mobily cell by criterion Active nodes
+      // with mobiles nodes means that the cell is part of this layer
+      for (auto dof_index : local_dofs_indices)
```

Comment:
Is this last loop necessary? I think it could be part of the last one.

Reply:
No I can't because I need to know the value of every nodes after adding the additional layer of mobile to know what cells are active. Otherwise I can assign active cells from mobile node from the 2nd loop.